solver/errdefs/context.go

```diff
@@ -3,11 +3,25 @@ package errdefs
 import (
 	"context"
 	"errors"
+	"strings"
 
 	"github.com/moby/buildkit/util/grpcerrors"
 	"google.golang.org/grpc/codes"
 )
 
-func IsCanceled(err error) bool {
-	return errors.Is(err, context.Canceled) || grpcerrors.Code(err) == codes.Canceled
+func IsCanceled(ctx context.Context, err error) bool {
+	if errors.Is(err, context.Canceled) || grpcerrors.Code(err) == codes.Canceled {
+		return true
+	}
+	// grpc does not set cancel correctly when stream gets cancelled and then Recv is called
+	if err != nil && ctx.Err() == context.Canceled {
+		// when this error comes from containerd it is not typed at all, just concatenated string
+		if strings.Contains(err.Error(), "EOF") {
```

Comment:
Doesn’t seem robust.
Where is the source of this EOF

Reply:
At least one of the places should be https://github.com/containerd/containerd/blob/main/content/proxy/content_writer.go#L44 but there might be more.

Should we ever match an error that is not a real cancellation error here it isn't a big issue. We have already verified that our own context was canceled so at least it was the user's intent. Matching non-cancel error in here means that the error does not get cached as we don't consider it a real error and another parallel process requesting the same result would need to try to recompute the result again.